server/src/util/my-repos.ts

```diff
@@ -0,0 +1,150 @@
+}
+
+/**
+ * Given a repo and a user, returns a weighted contribution score for the repo
+ * (0-100).
+ */
+function getContributionWeight(
+    github: GitHub, org: string, repo: string, user: string, userToken: string):
+    Promise<number> {
+  return new Promise(async (resolve, reject) => {
+    try {
+      // GitHub doesn't provide a v4 API equivalent for stats, so v3 API must be
+      // used.
+      const queryPath = `repos/${org}/${repo}/stats/contributors`;
+
+      let response = await github.get(queryPath, userToken, false);
+      let retries = 0;
+      // GitHub's API may serve a cached response and begin an asynchronous
+      // job to calculate required data.
+      while (response.statusCode !== 200 && retries++ < MAX_STATS_RETRIES) {
+        await sleep(1000 * retries);
+        response = await github.get(queryPath, userToken, false);
+      }
+
+      if (response.statusCode === 200) {
+        const stats = JSON.parse(response.body) as GitHubStatsResponse;
+        // Find contribution stats for the specified user.
+        const stat = stats.filter((x) => x.author.login === user);
+        if (stat.length === 0) {
+          resolve(0);
+          return;
+        }
+
+        let contributions = 0;
+        for (const week of stat[0].weeks) {
+          // Convert UNIX timestamp in seconds to Date object.
+          const date = new Date(Number(week.w) * 1000);
+          const timeSince = Math.max(
+                                (new Date().getTime() - date.getTime()),
+                                CONTRIBUTION_WINDOW) /
+              CONTRIBUTION_WINDOW;
+          // Produce a weight between 0-1 based on the weeks passed. Anything
+          // less than CONTRIBUTION_WINDOW will be ~1.
+          const timeWeight = Math.tanh(5 / timeSince);
+          // Weight any contribution by recency. Delta sizes are ignored.
+          if (week.a + week.d + week.c) {
+            contributions += timeWeight;
+          }
+        }
+
+        const score = 100 * Math.tanh(contributions / CONTRIBUTION_SIZE);
+        resolve(Math.round(score * 100) / 100);
+      } else {
+        reject(`Unable to fetch contribution stats for ${repo}. Got ${
+            response.statusCode} from GitHub.`);
+      }
+    } catch (err) {
+      reject(err);
+    }
+  });
+}
+
+const reposQuery = gql`
+  query MyRepos($login:String!, $cursor: String) {
+    user(login: $login) {
+      contributedRepositories(first: 100, after: $cursor) {
+        pageInfo {
+          endCursor
+          hasNextPage
+        }
+        nodes {
+          name
+          owner {
+            login
+          }
+        }
+      }
+    }
+  }
+`;
+
+// Based on API response documented at
+// https://developer.github.com/v3/repos/statistics/#get-contributors-list-with-additions-deletions-and-commit-counts
+type GitHubStatsResponse = [{
+  author: {login: string},
+  total: number,
+  weeks: [{w: string, a: number, d: number, c: number}],
```

Comment:
It's at the link, but would be nice to add a comment here too: (w)eek unix timestamp, (a)dditions, (d)eletions, (c)ommits

Reply:
Done.